src/feed.sol

```diff
@@ -0,0 +1,144 @@
+        // gas optimized initialization of writeOffs
+        // write off are hardcoded in the contract instead of init function params
+
+        // risk group recoveryRatePD
+        recoveryRatePD[0] = ONE;
+        recoveryRatePD[1] = 90 * 10**25;
+
+        // 60% -> 40% write off
+        writeOffs[0] = WriteOff(91, 6 * 10**26);
+        // 80% -> 20% write off
+        writeOffs[1] = WriteOff(90, 8 * 10**26);
+    }
+
+    function uniqueDayTimestamp(uint timestamp) public pure returns (uint) {
+        return (1 days) * (timestamp/(1 days));
+    }
+
+    /// maturityDate is a unix timestamp
+    function file(bytes32 what, bytes32 nftID_, uint maturityDate_) public {
+        if (what == "maturityDate") {
+            maturityDate[nftID_] = uniqueDayTimestamp(maturityDate_);
+        } else { revert("unknown config parameter");}
+    }
+
+    /// Ceiling Implementation
+    function borrow(uint loan, uint amount) external auth {
+        uint normalizedDay = uniqueDayTimestamp(now);
+
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+
+        bytes32 nftID_ = nftID(loan);
+        uint maturityDate_ = maturityDate[nftID_];
+
+        // calculate future value FV
+        uint fv = rmul(rmul(rpow(pile.loanRates(loan),  safeSub(maturityDate_, normalizedDay), ONE), amount), recoveryRatePD[risk[nftID(loan)]]);
+
+        futureValue[loan] = safeAdd(futureValue[loan], fv);
+
+        if (buckets[maturityDate_].value == 0) {
+            addBucket(maturityDate_, fv);
+            return;
+        }
+
+        buckets[maturityDate_].value = safeAdd(buckets[maturityDate_].value, fv);
+    }
+
+    /// adds a new bucket to the linked-list
+    function repay(uint loan, uint amount) external auth {
+        uint maturityDate_ = maturityDate[nftID(loan)];
+
+        // todo reduce bucket with loan futureValue relative to amount
+        buckets[maturityDate_].value = safeSub(buckets[maturityDate_].value, amount);
+
+        if (buckets[maturityDate_].value == 0) {
+            removeBucket(maturityDate_);
+        }
+    }
+
+    function calcDiscount() public view returns(uint) {
+        uint normalizedDay = uniqueDayTimestamp(now);
+        uint sum = 0;
+
+        uint currDate = normalizedDay;
+
+        if (currDate > lastBucket) {
+            return 0;
+        }
+
+        while(buckets[currDate].next == 0) { currDate = currDate + 1 days; }
+
+        while(currDate != NullDate)
+        {
+            sum = safeAdd(sum, rdiv(buckets[currDate].value, rpow(discountRate, safeSub(currDate, normalizedDay), ONE)));
+            currDate = buckets[currDate].next;
+        }
+        return sum;
+    }
+
+    /// returns the NAV (net asset value) of the pool
+    function currentNAV() public returns(uint) {
+        uint nav_ = calcDiscount();
+
+        // add write offs to NAV
+        for (uint i = 0; i < writeOffs.length; i++) {
```

Comment:
maybe add a comment somewhere that the loan needs to be pushed to the writeoff rate groups manually

Reply:
yes, we anyway will add additional comments